70_ima_policy.md

```diff
@@ -0,0 +1,348 @@
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place
+- [ ] User-facing documentation has been created in [keylime/keylime-docs]
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+## Summary
+
+This enhancement proposal changes and unifies the IMA configuration options by
+formally specifying a JSON schema. The tenant and verifier are modified to use
+the IMA policy instead of the old flat file format to simplify the process and
+allow further development like server side signature validation.
+
+For users of the old flat file format a upgrade path using a commandline tool
+will be provided.
+
+## Motivation
+
+The configuration for IMA validation is currently split over multiple options
+and is not well documented. This makes it confusing to use and hard to implement
+features like signature validation or storing all the information in a separate
+DB table to reuse.
+
+### Goals
+
+* Merge the current allow and exclude list into one single JSON object
+* Move `ima_sign_verification_keys` also into the JSON object
+* Deprecate flat file allow and exclude list
+* Simplify IMA related options in the tenant
+* Provide a tool for converting flat file allow and exclude list to JSON object
+
+### Non-Goals
+
+* Change the current behavior of the IMA validation
+* Change how the measured boot reference state works
+* Implement server side signature validation
+
+## Proposal
+
+The entries `ima_sign_verification_keys` and `allowlist` will be replaced with
+`ima_policy` in the verifier API. `ima_policy` takes a JSON object that contains
+all the required data for IMA validation. A schema for the JSON object will be
+provided in the Keylime repository.
+
+The `allowlist*` options in the tenant will be renamed to `ima_policy*`.
+
+The flat file format will be deprecated in favor of the JSON format. This is
+done to simplify future development like server side signature validation. For
+users of the old format a conversion tool will be provided.
+
+
+### User Stories
+
+#### Story 1a - Using the legacy flat file format
+* User generates a allowlist `ima_allowlist.txt` and a exclude list `exclude.txt`
+* Conversion to new JSON format with `keylime_convert_allowlist --allowlist ima_allowlist.txt --exclude exclude.txt --out ima_policy.json`
+* The conversion tool produces `ima_policy.json`
+* Adds a new agent with `keylime_tenant -c add --ima_policy ima.json ...`
+* Agent is added to attestation.
+
+#### Story 1b - Using the legacy flat file format with signature
+* User generates a allowlist `ima_allowlist.txt` and a exclude list
+  `exclude.txt`, allowlist signature `ima_allowlist.txt.sig` and the public key `allowlist.key`
+* Conversion to new JSON format with `keylime_convert_allowlist --allowlist ima_allowlist.txt --exclude exclude.txt --allowlist-sig ima_allowlist.txt.sig --allowlist-sig allowlist.key --out ima_policy.json`
```

Comment:
I guess the conversion tool is supposed to handle `ima_sign_verification_keys*` options too.

Reply:
Is noted in the implementation details.